gui/ConfigMain.cpp

```diff
@@ -0,0 +1,549 @@
+    } else {
+        removeIMButton->setEnabled(true);
+        // configureButton->setEnabled(true);
+        if (currentIMView->currentIndex().row() == 0) {
+            moveUpButton->setEnabled(false);
+        } else {
+            moveUpButton->setEnabled(true);
+        }
+        if (currentIMView->currentIndex().row() ==
+            currentIMView->model()->rowCount() - 1) {
+            moveDownButton->setEnabled(false);
+        } else {
+            moveDownButton->setEnabled(true);
+        }
+    }
+}
+// end of SLOTs
+
+QString ConfigMain::icon() { return "fcitx5-rime"; }
+
+QString ConfigMain::title() { return _("Fcitx5 Rime Config Tool"); }
+
+void ConfigMain::load() {
+    if(inError) {
+        return;
+    }
+    modelToUi();
+}
+
+void ConfigMain::disableUi(const char *message) {
+    setEnabled(false);
+    overlay->enable(message);
+    inError = true;
+}
+
+void ConfigMain::setModelFromLayout(QList<fcitx::Key> &model_keys,
+                                    QLayout *layout) {
+    fcitx::FcitxQtKeySequenceWidget * keywidget = getKeyWidgetFromLayout(layout);
+    Q_ASSERT(keywidget != NULL);
+    model_keys = keywidget->keySequence();
+}
+
+void ConfigMain::uiToModel() {
+    model->candidatePerWord_ = cand_cnt_spinbox->value();
+
+    setModelFromLayout(model->toggleKeys_, horizontallayout_toggle);
+    setModelFromLayout(model->asciiKeys_, horizontallayout_ascii);
+    setModelFromLayout(model->pgdownKeys_, horizontallayout_pagedown);
+    setModelFromLayout(model->pgupKeys_, horizontallayout_pageup);
+    setModelFromLayout(model->trasimKeys_, horizontallayout_trasim);
+    setModelFromLayout(model->halffullKeys_, horizontallayout_hfshape);
+
+    if (model->switchKeys_.size() >= 2) {
+        model->switchKeys_[0] = textToSwitchKey(shift_l_combo->currentIndex());
+        model->switchKeys_[1] = textToSwitchKey(shift_r_combo->currentIndex());
+    }
+
+    // clear cuurent model and save from the ui
+    for (int i = 0; i < model->schemas_.size(); i++) {
+        model->schemas_[i].index = 0;
+        model->schemas_[i].active = false;
+    }
+    QStandardItemModel *qmodel =
+        static_cast<QStandardItemModel *>(currentIMView->model());
+    QModelIndex parent;
+    int seqno = 1;
+    for (int r = 0; r < qmodel->rowCount(parent); ++r) {
+        QModelIndex index = qmodel->index(r, 0, parent);
+        QVariant name = qmodel->data(index);
+        for (int i = 0; i < model->schemas_.size(); i++) {
+            if (model->schemas_[i].name == name) {
+                model->schemas_[i].index = seqno++;
+                model->schemas_[i].active = true;
+            }
+        }
+    }
+    model->sortSchemas();
+}
+
+void ConfigMain::save() {
+    if(inError) {
+        return;
+    }
+    uiToModel();
+    QFutureWatcher<void> *futureWatcher = new QFutureWatcher<void>(this);
+    futureWatcher->setFuture(
+        QtConcurrent::run<void>(this, &ConfigMain::modelToYaml));
```

Comment:
this is run is new thread? Is it thread safe?

Reply:
Yes, it is a new thread. When the new thread is running, the UI will be temporarily disabled so it is thread-safe.

Reply:
UI disabled / thread safe is two irrelevant thing. I doubt if you can call RIME API from a difference thread because rime is also multi-threaded internally.